src/api/regex.c

```diff
@@ -84,18 +208,154 @@ static int f_pcre_match(lua_State *L) {
     we just detect this case and give up. */
     luaL_error(L, "regex matching error: \\K was used in an assertion to "
     " set the match start after its end");
+    if (regex_compiled) pcre2_code_free(re);
     pcre2_match_data_free(md);
     return 0;
   }
   for (int i = 0; i < rc*2; i++)
     lua_pushinteger(L, ovector[i]+offset+1);
+  if (regex_compiled) pcre2_code_free(re);
   pcre2_match_data_free(md);
   return rc*2;
 }
 
+static int f_pcre_gmatch(lua_State *L) {
+  /* pattern param */
+  bool regex_compiled = false;
+  pcre2_code* re = regex_get_pattern(L, &regex_compiled);
+  if (!re) return 0;
+  size_t subject_len = 0;
+
+  /* subject param */
+  const char* subject = luaL_checklstring(L, 2, &subject_len);
+
+  /* offset param */
+  size_t offset = regex_offset_relative(
+    luaL_optnumber(L, 3, 1), subject_len
+  ) - 1;
+
+  /* keep strings on closure to avoid being collected */
+  lua_settop(L, 2);
+
+  RegexState *state;
+  state = (RegexState*)lua_newuserdatauv(L, sizeof(RegexState), 0);
+
+  state->re = re;
+  state->match_data = pcre2_match_data_create_from_pattern(re, NULL);
+  state->subject = subject;
+  state->subject_len = subject_len;
+  state->offset = offset;
+  state->found = true;
+  state->regex_compiled = regex_compiled;
+
+  lua_pushcclosure(L, regex_gmatch_iterator, 3);
+  return 1;
+}
+
+static int f_pcre_gsub(lua_State *L) {
+  size_t subject_len = 0, replacement_len = 0;
+
+  bool regex_compiled = false;
+  pcre2_code* re = regex_get_pattern(L, &regex_compiled);
+  if (!re) return 0 ;
+
+  char* subject = (char*) luaL_checklstring(L, 2, &subject_len);
+  const char* replacement = luaL_checklstring(L, 3, &replacement_len);
+  const int limit = luaL_optinteger(L, 4, 0);
```

Comment:
Might be a good idea to check for negative values, so they don't cause the slow path to be taken.

Reply:
done